Packages/Tracking/Core/Runtime/Scripts/XRHandsSubsystem/XRHandsLeapProvider.cs

```diff
@@ -0,0 +1,499 @@
+            hand.StabilizedPalmPosition = hand.PalmPosition;
+            hand.PalmNormal = hand.Rotation * Vector3.down;
+            hand.Direction = hand.Rotation * Vector3.forward;
+
+            // Calculate now we have the hand data available.
+            // Requires `Hand.Rotation` and fingers to be set.
+            hand.GrabStrength = CalculateGrabStrength(ref hand);
+            hand.PinchStrength = CalculatePinchStrength(ref hand, handScale);
+            hand.PinchDistance = CalculatePinchDistance(ref hand);
+
+            // Other hand-properties are derived.
+            hand.PalmNormal = hand.Rotation * Vector3.down;
+
+            // Fill arm data.
+            var palmPosition = palmPose.position;
+            var wristPosition = wristPose.position;
+            var wristWidth = hand.PalmWidth * 0.6f;
+
+            const float elbowLength = 0.3f;
+            var elbowRotation = palmPose.rotation;
+            var elbowDirection = elbowRotation * Vector3.back;
+            var elbowPosition = palmPose.position + (elbowDirection * elbowLength);
+            var centerPosition = (elbowPosition + palmPosition) / 2f;
+            hand.Arm.Fill(
+                elbowPosition,
+                wristPosition,
+                centerPosition,
+                elbowDirection,
+                elbowLength,
+                wristWidth,
+                elbowRotation
+            );
+        }
+
+        float GetHandTimeVisible(Chirality chirality)
+        {
+            long currentDateTimeTicks = DateTime.Now.Ticks;
+
+            float timeVisible = 0;
+            if (chirality == Chirality.Left)
+            {
+                if (_leftHandFirstSeen_ticks == -1)
+                {
+                    _leftHandFirstSeen_ticks = currentDateTimeTicks;
+                    _leftHandId = _handId++;
+                }
+                timeVisible = ((float)(currentDateTimeTicks - _leftHandFirstSeen_ticks)) / (float)TimeSpan.TicksPerSecond;
+            }
+            else
+            {
+                if (_rightHandFirstSeen_ticks == -1)
+                {
+                    _rightHandFirstSeen_ticks = currentDateTimeTicks;
+                    _rightHandId = _handId++;
+                }
+                timeVisible = ((float)(currentDateTimeTicks - _rightHandFirstSeen_ticks)) /
+                              (float)TimeSpan.TicksPerSecond;
+            }
+
+            return timeVisible;
+        }
+
+        private float CalculateHandScale(ref Hand hand)
+        {
+            // Iterate through the fingers, skipping the thumb and accumulate the scale.
+            float scale = 0.0f;
+            for (var i = 1; i < hand.Fingers.Count; ++i)
+            {
+                scale += (hand.Fingers[i].Bone(Bone.BoneType.TYPE_METACARPAL).Length / DefaultMetacarpalLengths[i]) / 4.0f;
+            }
+
+            return scale;
+        }
+
+        private float CalculatePinchStrength(ref Hand hand, float handScale)
```

Comment:
Is some of this code duplicated elsewhere in the plugin? If so would be good to share the common code?